# Preprocessing Training and Modeling

In [1]:
import numpy as np
import pandas as pd
import math
import seaborn as sns
import matplotlib.pyplot as plt
import datetime

## Preprocessing and Understanding

In [2]:
df = pd.read_csv('fraudTrain.csv')

In [3]:
# Deleted first column since it was the same as the index
df = df.iloc[: , 1:]
# We are going to rename some columns for ease
df = df.rename(columns={'trans_date_trans_time':'date_time'})
# We want to change the data type of the 'date_time' column to datetime data type for easier manipulation of data
df['date_time'] = pd.to_datetime(df['date_time'])
df['date_time'].dtypes
# Drop any duplicates
df.drop_duplicates(inplace=True)

In [4]:
df.shape

(1296675, 22)

In [5]:
# Let's downsize the data for ease of speed
df = df.sample(frac=0.01)
df.shape

(12967, 22)

In [6]:
df.head()

date_time            cc_num  \
500767  2019-08-05 14:34:59  6506982560413523   
784441  2019-12-01 03:09:14      502038091380   
519637  2019-08-12 08:38:40   180049032966888   
1150334 2020-04-24 21:18:06  4452366298769043   
348021  2019-06-12 17:41:31  2288813824604479   

                                   merchant       category     amt    first  \
500767                    fraud_Beier-Hyatt   shopping_pos    2.18    Alice   
784441                     fraud_Barton Inc    grocery_pos   93.66     Kurt   
519637      fraud_Little, Gutmann and Lynch   shopping_net   99.82  Michael   
1150334  fraud_Turner, Ruecker and Parisian       misc_pos  317.19    Linda   
348021                fraud_Gottlieb-Hansen  personal_care   20.66  Barbara   

           last gender                              street            city  \
500767     Kemp      F                 057 Martinez Radial       Texarkana   
784441   Peters      M                    7290 Ashlee Keys    Kissee Mills   
519637   Flores      M  70761 Fitzpatrick Brooks Suite 631           Saxon   
1150334   Davis      F            6602 Ortiz Pine Apt. 179  Blooming Grove   
348021   Norman      F               6278 Stephanie Unions   New York City   

         ...      lat     long  city_pop                             job  \
500767   ...  33.4310 -93.8765     36438                         Surgeon   
784441   ...  36.6704 -93.0377      1201          Engineer, agricultural   
519637   ...  46.4959 -90.4383       795  Television/film/video producer   
1150334  ...  32.0758 -96.7010      1563               Financial adviser   
348021   ...  40.8265 -73.9383   1577385                       Herbalist   

                dob                         trans_num   unix_time  merch_lat  \
500767   1985-04-15  fe78243a04b93f77f72271f49e178460  1344177299  34.179876   
784441   1980-03-18  a0063670117d27d3de3ac3a7ba279e9d  1354331354  36.774243   
519637   1986-04-15  5dfbfb5b5b3a357a5fef0d599a2647d0  1344760720  46.296291   
1150334  1978-03-04  8122ee0845b03f9aa5fc162530320432  1366838286  31.369695   
348021   1981-08-29  1edbd647c515874abbcf8bdf051cab61  1339522891  40.795240   

         merch_long  is_fraud  
500767   -93.768268         0  
784441   -93.523791         0  
519637   -89.489390         0  
1150334  -96.949437         0  
348021   -74.840796         0  

[5 rows x 22 columns]

In [7]:
df.columns

Index(['date_time', 'cc_num', 'merchant', 'category', 'amt', 'first', 'last',
       'gender', 'street', 'city', 'state', 'zip', 'lat', 'long', 'city_pop',
       'job', 'dob', 'trans_num', 'unix_time', 'merch_lat', 'merch_long',
       'is_fraud'],
      dtype='object')

In [8]:
# No null data
df.isnull().sum()

date_time     0
cc_num        0
merchant      0
category      0
amt           0
first         0
last          0
gender        0
street        0
city          0
state         0
zip           0
lat           0
long          0
city_pop      0
job           0
dob           0
trans_num     0
unix_time     0
merch_lat     0
merch_long    0
is_fraud      0
dtype: int64

In [9]:
print(df['amt'].max())
print(df['amt'].min())

15305.95
1.0


Since there is a huge difference in the max and min values for the transaction amount, I will then scale this column to make the data more practical.

In [10]:
# We want to standardize the 'amt' column
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
amount = df['amt'].values
df['amt'] = sc.fit_transform(amount.reshape(-1, 1))
df['amt']

500767    -0.376352
784441     0.133939
519637     0.168301
1150334    1.380829
348021    -0.273268
             ...   
8395      -0.093539
1249845   -0.181562
848046    -0.055775
260209     0.451003
1200601   -0.128068
Name: amt, Length: 12967, dtype: float64

Everything in the 'amt' or Amount column looks like it has been scaled and so we will move onto the train and test split. While doing EDA, we found out that most of the columns are not correlated with wether or not the transaction is fraudulent. Therefore, we will create a new dataframe that focuses on the features we need. Additionally, it will save running time.

In [11]:
dfa = df[['amt', 'is_fraud']]
dft = df[['date_time', 'is_fraud']]

## Train Test Split - Amount

In [12]:
X = dfa.drop('is_fraud', axis = 1).values
y = dfa['is_fraud'].values

In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 1)

### Decision Tree Classifier

In [14]:
from sklearn.tree import DecisionTreeClassifier

DT = DecisionTreeClassifier(max_depth = 4, criterion = 'entropy')
DT.fit(X_train, y_train)
dt_yhat = DT.predict(X_test)

In [15]:
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

print('Accuracy score of the Decision Tree model is {}'.format(accuracy_score(y_test, dt_yhat)))
print('F1 score of the Decision Tree model is {}'.format(f1_score(y_test, dt_yhat)))
confusion_matrix(y_test, dt_yhat, labels = [0, 1])

Accuracy score of the Decision Tree model is 0.9938309685379395
F1 score of the Decision Tree model is 0.23076923076923078


array([[3219,    0],
       [  20,    3]], dtype=int64)

Here, the first row represents positive and the second row represents negative. So, we have 3219 as true positive and 0 are false positive. We have 3219 that are successfully classified as a nonfraudulent transaction and 0 were falsely classified as nonfraudulent, but they were fraudulent. For the accuracy and F1 score, generally we want the score closer to 1. The accuracy score of the Decision Tree model is pretty good but the F1 score is alright.

### K-Nearest Neighbors

In [21]:
from sklearn.neighbors import KNeighborsClassifier

n = 7
KNN = KNeighborsClassifier(n_neighbors = n)
KNN.fit(X_train, y_train)
knn_yhat = KNN.predict(X_test)

In [22]:
print('Accuracy score of the K-Nearest Neighbors model is {}'.format(accuracy_score(y_test, knn_yhat)))
print('F1 score of the K-Nearest Neighbors model is {}'.format(f1_score(y_test, knn_yhat)))
confusion_matrix(y_test, knn_yhat, labels = [0, 1])

Accuracy score of the K-Nearest Neighbors model is 0.9935225169648365
F1 score of the K-Nearest Neighbors model is 0.16


array([[3219,    0],
       [  21,    2]], dtype=int64)

#### Hyperparameter Tuning

In [23]:
from sklearn.model_selection import GridSearchCV

# List Hyperparameters that we want to tune
KNN_2 = KNeighborsClassifier()
k_range = list(range(1, 31))
print(k_range)
param_grid = dict(n_neighbors=k_range)
print(param_grid)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]
{'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]}


### Logistic Regression

In [24]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()
lr.fit(X_train, y_train)
lr_yhat = lr.predict(X_test)

In [25]:
print('Accuracy score of the Logistic Regression model is {}'.format(accuracy_score(y_test, lr_yhat)))
print('F1 score of the Logistic Regression model is {}'.format(f1_score(y_test, lr_yhat)))
confusion_matrix(y_test, lr_yhat, labels = [0, 1])

Accuracy score of the Logistic Regression model is 0.9919802590993214
F1 score of the Logistic Regression model is 0.0


array([[3216,    3],
       [  23,    0]], dtype=int64)

### Random Forest

In [26]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(max_depth = 4)
rf.fit(X_train, y_train)
rf_yhat = rf.predict(X_test)

In [27]:
print('Accuracy score of the Random Forest model is {}'.format(accuracy_score(y_test, rf_yhat)))
print('F1 score of the Random Forest model is {}'.format(f1_score(y_test, rf_yhat)))
confusion_matrix(y_test, rf_yhat, labels = [0, 1])

Accuracy score of the Random Forest model is 0.9935225169648365
F1 score of the Random Forest model is 0.22222222222222218


array([[3218,    1],
       [  20,    3]], dtype=int64)

### XGBoost

In [28]:
from xgboost import XGBClassifier

xgb = XGBClassifier(max_depth = 4)
xgb.fit(X_train, y_train)
xgb_yhat = xgb.predict(X_test)

In [29]:
print('Accuracy score of the XGBoost model is {}'.format(accuracy_score(y_test, xgb_yhat)))
print('F1 score of the XGBoost model is {}'.format(f1_score(y_test, xgb_yhat)))
confusion_matrix(y_test, xgb_yhat, labels = [0, 1])

Accuracy score of the XGBoost model is 0.9935225169648365
F1 score of the XGBoost model is 0.3225806451612903


array([[3216,    3],
       [  18,    5]], dtype=int64)

### Hybrid XGBoost Model - XGBoost and Random Forest Classifier

In [30]:
from xgboost import XGBRFClassifier

xgbrf = XGBRFClassifier()
xgbrf.fit(X_train, y_train)
xgbrf_yhat = xgb.predict(X_test)

In [31]:
print('Accuracy score of the XGBRF model is {}'.format(accuracy_score(y_test, xgbrf_yhat)))
print('F1 score of the XGBRF model is {}'.format(f1_score(y_test, xgbrf_yhat)))
confusion_matrix(y_test, xgbrf_yhat, labels = [0, 1])

Accuracy score of the XGBRF model is 0.9935225169648365
F1 score of the XGBRF model is 0.3225806451612903


array([[3216,    3],
       [  18,    5]], dtype=int64)

## Overall Scores

In [34]:
accuracy = [accuracy_score(y_test, dt_yhat), accuracy_score(y_test, knn_yhat), accuracy_score(y_test, lr_yhat), 
            accuracy_score(y_test, xgb_yhat), accuracy_score(y_test, xgbrf_yhat)]
f1 = [f1_score(y_test, dt_yhat), f1_score(y_test, knn_yhat), f1_score(y_test, lr_yhat), 
            f1_score(y_test, xgb_yhat), f1_score(y_test, xgbrf_yhat)]
names = ['DT', 'KNN', 'LR', 'XGB', 'XBGRF']

In [64]:
def scores(accuracy, f1, names):
    for i in range(0, 5):
        print('For', names[i])
        print('It has an accuracy score of', accuracy[i])
        print('And an F1 score of', f1[i])

In [65]:
scores(accuracy, f1, names)

For DT
It has an accuracy score of 0.9938309685379395
And an F1 score of 0.23076923076923078
For KNN
It has an accuracy score of 0.9935225169648365
And an F1 score of 0.16
For LR
It has an accuracy score of 0.9919802590993214
And an F1 score of 0.0
For XGB
It has an accuracy score of 0.9935225169648365
And an F1 score of 0.3225806451612903
For XBGRF
It has an accuracy score of 0.9935225169648365
And an F1 score of 0.3225806451612903


In [66]:
max(accuracy)

0.9938309685379395

In [67]:
max(f1)

0.3225806451612903

Decision Tree Classifier has the highest accuracy score but XGBoost and XGBoost/Random Forest has the highest F1 score. So how do we choose between the two?

## Conclusion

Remember that the F1 score is balancing precision and recall on the positive class while accuracy looks at correctly classified observations both positive and negative. That makes a big difference especially for the imbalanced problems where by default our model will be good at predicting true negatives and hence accuracy will be high. However, if you care equally about true negatives and true positives then accuracy is the metric you should choose. 

Since our dataset is naturally imbalanced, it will make our accuracy score really high by default. That is why we will care more about the F1 score in this case.

Therefore, the best model is the XGBoost and XGBRF which has the same accuracy and F1 score. We have an accuracy of 99.35% and a F1 score of 32.26%